applications/StructuralMechanicsApplication/custom_elements/timoshenko_beam_element_2D2N.cpp

```diff
@@ -0,0 +1,1051 @@
+        StructuralMechanicsElementUtilities::BuildRotationMatrixFor2D2NBeam(T, angle);
+        noalias(Tt) = trans(T);
+
+        // We rotate each submatrix independently
+        RangeMatrixType rLHSaa(rLHS, boost::numeric::ublas::range(0, 3), boost::numeric::ublas::range(0, 3));
+        RangeMatrixType rLHSab(rLHS, boost::numeric::ublas::range(0, 3), boost::numeric::ublas::range(3, 6));
+        RangeMatrixType rLHSba(rLHS, boost::numeric::ublas::range(3, 6), boost::numeric::ublas::range(0, 3));
+        RangeMatrixType rLHSbb(rLHS, boost::numeric::ublas::range(3, 6), boost::numeric::ublas::range(3, 6));
+
+        rLHSaa = prod(T, MatrixType(prod(rLHSaa, Tt)));
+        rLHSab = prod(T, MatrixType(prod(rLHSab, Tt)));
+        rLHSba = prod(T, MatrixType(prod(rLHSba, Tt)));
+        rLHSbb = prod(T, MatrixType(prod(rLHSbb, Tt)));
+
+        VectorType rRHSa(3);
+        rRHSa[0] = rRHS[0];
+        rRHSa[1] = rRHS[1];
+        rRHSa[2] = rRHS[2];
+        VectorType rRHSb(3);
+        rRHSb[0] = rRHS[3];
+        rRHSb[1] = rRHS[4];
+        rRHSb[2] = rRHS[5];
+
+        rRHSa = prod(T, rRHSa);
+        rRHSb = prod(T, rRHSb);
+
+        rRHS[0] = rRHSa[0];
+        rRHS[1] = rRHSa[1];
+        rRHS[2] = rRHSa[2];
+        rRHS[3] = rRHSb[0];
+        rRHS[4] = rRHSb[1];
+        rRHS[5] = rRHSb[2];
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+void TimoshenkoBeamElement2D2N::CalculateOnIntegrationPoints(
+    const Variable<double>& rVariable,
+    std::vector<double>& rOutput,
+    const ProcessInfo& rProcessInfo
+    )
+{
+    const auto& integration_points = IntegrationPoints(GetIntegrationMethod());
+    rOutput.resize(integration_points.size());
+
+    if (rVariable == AXIAL_FORCE) {
+
+        const auto &r_geometry = GetGeometry();
+
+        ConstitutiveLaw::Parameters cl_values(r_geometry, GetProperties(), rProcessInfo);
+        auto &r_cl_options = cl_values.GetOptions();
+        r_cl_options.Set(ConstitutiveLaw::COMPUTE_STRESS             , true);
+        r_cl_options.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, false);
+
+        const double Phi    = CalculatePhi(cl_values);
+        const double length = StructuralMechanicsElementUtilities::CalculateReferenceLength2D2N(*this);
+
+        // Let's initialize the cl values
+        VectorType strain_vector(3), stress_vector(3);
+        strain_vector.clear();
+        cl_values.SetStrainVector(strain_vector);
+        cl_values.SetStressVector(stress_vector);
+        VectorType nodal_values(6);
+        GetNodalValuesVector(nodal_values);
+
+        // Loop over the integration points
+        for (SizeType IP = 0; IP < integration_points.size(); ++IP) {
+            const double xi = integration_points[IP].X();
+
+            strain_vector[0] = CalculateAxialStrain(length, Phi, xi, nodal_values);      // El
```

Comment:
i would suggest a function "ComputeGeneralizedStrain which computes directly the stain vector

Reply:
done